### PR TITLE
[TableGen][InstrInfoEmitter] Count sub-operands on def operands

### PR DESCRIPTION
**Pull request by darkbuck**
- If a def operand includes multiple sub-operands, count them when
  generating instr info.
- Found issues in x86 and sparc backends, where memory operands of
  store or store-like instructions are wrongly placed in the output
  list.

### CODE REVIEW

File: llvm/test/TableGen/def-multiple-operands.td [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// RUN: llvm-tblgen -gen-instr-info -I %p/../../include %s | FileCheck %s
+
```
Comment (Pierre-vh) [inline llvm/test/TableGen/def-multiple-operands.td:2]: Can you add a short description of what this test is checking?

**Review by Pierre-vh — COMMENTED** [review]
I don't really understand the backend changes so I'll leave it up to other to review those

File: llvm/lib/Target/Sparc/SparcInstrInfo.td [hunk 2]
```diff
@@ -693,38 +693,38 @@ let DecoderNamespace = "SparcV8", Predicates = [HasNoV9] in {
 }
 
 let rd = 0 in {
-  let Defs = [CPSR] in {
-    def STCSRrr : F3_1<3, 0b110101, (outs (MEMrr $rs1, $rs2):$addr), (ins),
+  let mayStore = 1, Uses = [CPSR] in {
```
Comment (arsenm) [inline llvm/lib/Target/Sparc/SparcInstrInfo.td:696]: The mayStore setting looks like an unrelated change

File: llvm/lib/Target/X86/X86InstrCMovSetCC.td [hunk 3]
```diff
@@ -58,8 +58,8 @@ let SchedRW = [WriteCMOV.Folded, WriteCMOV.ReadAfterFold] in {
 }
 let SchedRW = [WriteCMOV, ReadDefault, ReadDefault, ReadDefault, ReadDefault, ReadDefault],
     Predicates = [HasCMOV, HasCF, In64BitMode], mayStore = 1 in
-  def mr : ITy<0x40, MRMDestMemCC, t, (outs t.MemOperand:$dst),
-                (ins t.RegClass:$src1, ccode:$cond),
+  def mr : ITy<0x40, MRMDestMemCC, t, (outs),
+                (ins t.MemOperand:$dst, t.RegClass:$src1, ccode:$cond),
```
Comment (XinWang10) [inline llvm/lib/Target/X86/X86InstrCMovSetCC.td:62]: This change looks good to me.

File: llvm/test/TableGen/def-multiple-operands.td (hunk 1 above)
Comment (darkbuck) [reply llvm/test/TableGen/def-multiple-operands.td:2]: Done, this test checks whether the `MCInstrDesc` entry for `InstA` has the expected `NumOperands` and `NumDefs`. It should have 3 defs out of 4 operands.

File: llvm/lib/Target/Sparc/SparcInstrInfo.td
[the remark this answers is not in the document]
Comment (darkbuck) [reply llvm/lib/Target/Sparc/SparcInstrInfo.td:1]: > Overall the changes looks okay to me, but a little question; why doesn't it also affect the normal store instructions?

These STORE-like instructions have the wrong operand info for their memory operands, which should be 'use' instead of 'def'. I found that during the debugging of this change.

File: llvm/lib/Target/Sparc/SparcInstrInfo.td (hunk 2 above)
Comment (darkbuck) [reply llvm/lib/Target/Sparc/SparcInstrInfo.td:696]: > The mayStore setting looks like an unrelated change

I agree but I don't want those instructions are optimized away after this change. I cannot find any test from sparc target. Not sure whom should we ask for help.

Comment (arsenm) [reply llvm/lib/Target/Sparc/SparcInstrInfo.td:696]: But can you precommit them? That should be doable without the operand change?

**Comment by darkbuck** [comment]
this commit was split into 3
